runtime/vara/src/tests.rs

```diff
@@ -20,15 +20,345 @@ use super::*;
-    check_instructions_weights, check_lazy_pages_costs, check_pages_costs, check_syscall_weights,
-    PagesCosts,
-};
 use sp_runtime::AccountId32;
 
+const INSTRUCTIONS_SPREAD: u8 = 50;
+const SYSCALL_SPREAD: u8 = 10;
+const PAGES_SPREAD: u8 = 10;
+
+/// Structure to hold weight expectation
+struct WeightExpectation {
+    weight: u64,
+    expected: u64,
+    spread: u8,
+    name: &'static str,
+}
+
+impl WeightExpectation {
+    fn new(weight: u64, expected: u64, spread: u8, name: &'static str) -> Self {
+        Self {
+            weight,
+            expected,
+            spread,
+            name,
+        }
+    }
+
+    fn check(&self) -> Result<(), String> {
+        let left = self.expected - self.expected * self.spread as u64 / 100;
+        let right = self.expected + self.expected * self.spread as u64 / 100;
+
+        if left > self.weight || self.weight > right {
+            return Err(format!("Instruction [{}]. Weight is {} ps. Expected weight is {} ps. {}% spread interval: [{left} ps, {right} ps]", self.name, self.weight, self.expected, self.spread));
+        }
+
+        Ok(())
+    }
+}
+
+fn check_expectations(expectations: &[WeightExpectation]) -> Result<(), Vec<String>> {
+    let errors = expectations
+        .iter()
+        .filter_map(|expectation| {
+            if let Err(err) = expectation.check() {
+                Some(err)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<String>>();
+
+    if errors.is_empty() {
+        Ok(())
+    } else {
+        Err(errors)
+    }
+}
+
+/// Memory pages access costs.
+struct PagesCosts {
+    pub load_page_data: CostOf<GearPagesAmount>,
+    pub upload_page_data: CostOf<GearPagesAmount>,
+    pub mem_grow: CostOf<GearPagesAmount>,
+    pub mem_grow_per_page: CostOf<GearPagesAmount>,
+    pub parachain_read_heuristic: CostOf<GearPagesAmount>,
+}
+
+impl<T: pallet_gear::Config> From<MemoryWeights<T>> for PagesCosts {
+    fn from(val: MemoryWeights<T>) -> Self {
+        Self {
+            load_page_data: val.load_page_data.ref_time().into(),
+            upload_page_data: val.upload_page_data.ref_time().into(),
+            mem_grow: val.mem_grow.ref_time().into(),
+            mem_grow_per_page: val.mem_grow_per_page.ref_time().into(),
+            parachain_read_heuristic: val.parachain_read_heuristic.ref_time().into(),
+        }
+    }
+}
+
+/// Check that the weights of instructions are within the expected range
+fn check_instructions_weights<T: pallet_gear::Config>(
+    weights: InstructionWeights<T>,
+    expected: InstructionWeights<T>,
+) -> Result<(), Vec<String>> {
+    macro_rules! expectation {
+        ($inst_name:ident) => {
+            WeightExpectation::new(
+                weights.$inst_name.into(),
+                expected.$inst_name.into(),
+                INSTRUCTIONS_SPREAD,
+                stringify!($inst_name),
+            )
+        };
+    }
+
+    let expectations = vec![
```

Comment:
It'd be great also to have some check of missing variants

Reply:
Yeah, added checks for the number of fields to be checked